pkg/providers/cloudstack/decoder/decoder.go

```diff
@@ -16,10 +17,52 @@ const (
 	EksaCloudStackHostPathToMount         = "EKSA_CLOUDSTACK_HOST_PATHS_TO_MOUNT"
 	defaultVerifySslValue                 = "true"
 	CloudStackGlobalAZ                    = "global"
+
+	APIKeyKey    = "api-key"
+	SecretKeyKey = "secret-key"
+	APIUrlKey    = "api-url"
+	VerifySslKey = "verify-ssl"
 )
 
-// ParseCloudStackSecret parses the input b64 string into the ini object to extract out the api key, secret key, and url.
-func ParseCloudStackSecret() (*CloudStackExecConfig, error) {
+// ParseCloudStackCredsFromSecrets parses a list of secrets to extract out the api keys, secret keys, and urls.
+func ParseCloudStackCredsFromSecrets(secrets []apiv1.Secret) (*CloudStackExecConfig, error) {
+	if len(secrets) == 0 {
+		return nil, fmt.Errorf("no secrets provided - unable to generate CloudStackExecConfig")
+	}
+	cloudstackProfiles := make([]CloudStackProfileConfig, len(secrets))
+	for idx, secret := range secrets {
+		apiKey, ok := secret.Data[APIKeyKey]
+		if !ok {
+			return nil, fmt.Errorf("secret %s is missing required key %s", secret.Name, APIKeyKey)
+		}
+		secretKey, ok := secret.Data[SecretKeyKey]
+		if !ok {
+			return nil, fmt.Errorf("secret %s is missing required key %s", secret.Name, SecretKeyKey)
+		}
+		apiURL, ok := secret.Data[APIUrlKey]
+		if !ok {
+			return nil, fmt.Errorf("secret %s is missing required key %s", secret.Name, APIUrlKey)
+		}
+		verifySsl, ok := secret.Data[VerifySslKey]
+		if !ok {
+			verifySsl = []byte(defaultVerifySslValue)
+		}
+		cloudstackProfiles[idx] = CloudStackProfileConfig{
+			Name:          secret.Name,
+			ApiKey:        string(apiKey),
+			SecretKey:     string(secretKey),
+			ManagementUrl: string(apiURL),
+			VerifySsl:     string(verifySsl),
+		}
+	}
```

Comment:
If you want to use append
```suggestion
	cloudstackProfiles := make([]CloudStackProfileConfig, 0, len(secrets))
	for idx, secret := range secrets {
		apiKey, ok := secret.Data[APIKeyKey]
		if !ok {
			return nil, fmt.Errorf("secret %s is missing required key %s", secret.Name, APIKeyKey)
		}
		secretKey, ok := secret.Data[SecretKeyKey]
		if !ok {
			return nil, fmt.Errorf("secret %s is missing required key %s", secret.Name, SecretKeyKey)
		}
		apiURL, ok := secret.Data[APIUrlKey]
		if !ok {
			return nil, fmt.Errorf("secret %s is missing required key %s", secret.Name, APIUrlKey)
		}
		verifySsl, ok := secret.Data[VerifySslKey]
		if !ok {
			verifySsl = []byte(defaultVerifySslValue)
		}
		cloudstackProfiles = append(
		    cloudstackProfiles,
		    CloudStackProfileConfig{
			   Name:          secret.Name,
			   ApiKey:        string(apiKey),
			   SecretKey:     string(secretKey),
			   ManagementUrl: string(apiURL),
			   VerifySsl:     string(verifySsl),
		   },
		)
	}
```

Reply:
Sounds good, thanks. Could you explain the value of using this approach over the original one?

Reply:
If with the original one you mean the one you had originally, the advantage is you avoid multiple array resizing (which involves memory allocation and memory copy) and your slice only takes the memory it needs (the auto sizing algorithm doesn't know how many elements you need so it might overshoot the size.

If you are asking compared to your current code, using `slice[position` assignment, I think what I propose is simpler (since you avoid caring about positioning), less prone to errors and completely removes the possibility of having an out of bounds panic bug.